examples/amm/lib.rs

```diff
@@ -0,0 +1,373 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+const PRECISION: u128 = 1_000_000;
+
+#[ink::contract]
+mod amm {
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        ZeroLiquidity,
+        ZeroAmount,
+        InsufficientAmount,
+        NonEquivalentAmount,
+        ThresholdNotReached,
+        InvalidShare,
+        InsufficientLiquidity,
+        SlippageExceeded,
+    }
+
+    #[derive(Default, SpreadAllocate)]
+    #[ink(storage)]
+    pub struct Amm {
+        total_shares: Balance,
+        total_token1: Balance,
+        total_token2: Balance,
+        token1_balance: Mapping<AccountId, Balance>,
+        token2_balance: Mapping<AccountId, Balance>,
+        lp_shares: Mapping<AccountId, Balance>,
+        fee: Balance,
+    }
+
+    #[ink(impl)]
+    impl Amm {
+        fn has_valid_amount(
+            &self,
+            balance: &Mapping<AccountId, Balance>,
+            _qty: Balance,
+        ) -> Result<(), Error> {
+            let caller = self.env().caller();
+            let shares = balance.get(&caller).unwrap_or(0);
+
+            match _qty {
+                0 => Err(Error::ZeroAmount),
+                _ if _qty > shares => Err(Error::InsufficientAmount),
+                _ => Ok(()),
+            }
+        }
+
+        fn constant_k(&self) -> Balance {
+            self.total_token1 * self.total_token2
+        }
+
+        fn has_active_pool(&self) -> Result<(), Error> {
+            match self.constant_k() {
+                0 => Err(Error::ZeroLiquidity),
+                _ => Ok(()),
+            }
+        }
+
+        #[ink(constructor)]
+        pub fn new(_fee: Balance) -> Self {
+            ink_lang::utils::initialize_contract(Self::new_init);
+
+            Self {
+                fee: if _fee >= 1000 { 0 } else { _fee },
+                ..Default::default()
+            }
+        }
+
+        fn new_init(&mut self) {}
```

Comment:
The reason your test fails is because of the incorrect usage of `initialize_contract`. Here you are returning a manually constructed instance of the contract, `Amm`, which initializes everything to its default value. This means all three `Mapping`'s in the `Amm` struct will have a default storage key of `0`.

```
[lib.rs:302] &contract = Amm {
    total_shares: 0,
    total_token1: 0,
    total_token2: 0,
    token1_balance: Mapping {
        offset_key: Key(0x_00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000),
    },
    token2_balance: Mapping {
        offset_key: Key(0x_00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000),
    },
    lp_shares: Mapping {
        offset_key: Key(0x_00000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000),
    },
    fee: 0,
}
```

What needs to be done instead is to use the instance of the contract returned by `initialize_contract`. You can see an example of how to do this correctly [here](https://ink.substrate.io/datastructures/mapping/).

 ```suggestion
        pub fn new(_fee: Balance) -> Self {
            ink_lang::utils::initialize_contract(|contract: &mut Self| {
                contract.fee = if _fee >= 1000 { 0 } else { _fee };
            })
        }
```

One more note: the examples are still using `alpha.1` - we should probably release `alpha.2` which removes the need for `SpreadAllocate` and `initialize_mapping` which would make this issue irrelevant